### PR TITLE
ADBDEV-6139 Fix clang compiler errors 

### PR DESCRIPTION
**Pull request by dnskvlnk**
Fix clang compiler errors

This patch fixes multiple errors produced by clang:
1. -Wimplicit-fallthrough was fixed by adding generic macro
2. -Wunused-function
3. -Wunused-variable and -Wunused-but-set-variable
4. -Wvla-cxx-extension
5. Unknown "asm" keyword
6. -Wdeprecated-declarations in third party headers was fixed by disabling this error for particular headers

Also, this patch adds an ability to build gpdb in a docker container using clang

### CODE REVIEW

File: src/backend/utils/error/elog.c [hunk 1]
```diff
@@ -167,11 +167,11 @@ static void write_eventlog(int level, const char *line, int len);
  * It only works for x86 processors.
  */
 #if defined(__i386)
-#define ASMFP asm volatile ("movl %%ebp, %0" : "=g" (ulp));
```
Comment (bandetto) [inline src/backend/utils/error/elog.c:170]: `asm` keyword is from Microsoft Visual C++. Are you sure this part is not for Windows?

File: gpcontrib/gpcloud/include/s3common_headers.h [hunk 2]
```diff
@@ -26,6 +26,23 @@ using std::string;
...
+		_Pragma("clang diagnostic pop") \
+	} while(0)
+#elif defined(__GNUC__)
+	#define SUPPRESS_COMPILER_WARNING(expr, warning) do { \
+		_Pragma("GCC diagnostic push") \
+		DO_PRAGMA(GCC diagnostic ignored warning) \
+		expr; \
+		_Pragma("GCC diagnostic pop") \
+	} while(0)
+#endif
```
Comment (bandetto) [inline gpcontrib/gpcloud/include/s3common_headers.h:44]: Why do we need to copy these macro definitions 2 times? Can't `c.h` be included here as well?

**Review by bandetto — COMMENTED** [review]
PR description:
- Add the PR title as the first line
- -Wunsued-variable -> -Wunused-variable
 - this patch added -> this patch adds

File: gpcontrib/gpcloud/include/s3common_headers.h [hunk 3]
```diff
@@ -26,6 +26,23 @@ using std::string;
...
 using std::vector;
 
+#define DO_PRAGMA(x) _Pragma (#x)
+#if defined(__clang__)
+	#define SUPPRESS_COMPILER_WARNING(expr, warning) do { \
+		_Pragma("clang diagnostic push") \
+		DO_PRAGMA(clang diagnostic ignored warning) \
+		expr; \
+		_Pragma("clang diagnostic pop") \
+	} while(0)
```
Comment (bandetto) [inline gpcontrib/gpcloud/include/s3common_headers.h:36]: Since it is only used with "-Wdeprecated-declarations", can this just be changed to something like this:
```suggestion
	#define SUPPRESS_DEPRECATION(expr) do { \
		_Pragma("clang diagnostic push") \
		DO_PRAGMA(clang diagnostic ignored "-Wdeprecated-declarations") \
		expr; \
		_Pragma("clang diagnostic pop") \
	} while(0)
```

Can this technique be used to push, include the header with deprecated functions, and then pop to avoid the need to use this macro for every call?

File: gpcontrib/gpcloud/include/s3common_headers.h (hunk 2 above)
Comment (dnskvlnk) [reply gpcontrib/gpcloud/include/s3common_headers.h:44]: s3 doesn't depend on postgres at all and I don't want to bring a new whole dependency because of one macro

File: gpcontrib/gpcloud/include/s3common_headers.h (hunk 3 above)
Comment (dnskvlnk) [reply gpcontrib/gpcloud/include/s3common_headers.h:36]: > Since it is only used with "-Wdeprecated-declarations", can this just be changed

No, more warnings in the future might appear and they won't necessarily be like this one. This macro was made as generic as possibly intentionally.

> Can this technique be used to push, include the header with deprecated functions, and then pop to avoid the need to use this macro for every call?

What do you mean? We either use a macro for every line which needs to be ignored or just ignore everything in a file. Why push-pop for each line? Also, I don't want to use push-pop semantics as one can forget to "pop" and ignore warnings which must not be ignored. One-liners make it more clear which expression should be ignored as well.

File: src/backend/utils/error/elog.c (hunk 1 above)
Comment (dnskvlnk) [reply src/backend/utils/error/elog.c:170]: [What do you mean?](https://en.cppreference.com/w/c/language/asm#:~:text=When%20compiling%20in%20ISO%20C%20mode%20by%20GCC%20or%20Clang)

File: gpcontrib/gpcloud/include/s3common_headers.h (hunk 3 above)
Comment (bandetto) [reply gpcontrib/gpcloud/include/s3common_headers.h:36]: > No, more warnings in the future might appear

Do we plan to introduce more warnings? :D

File: src/backend/utils/error/elog.c (hunk 1 above)
Comment (bandetto) [reply src/backend/utils/error/elog.c:170]: Nevermind.

File: gpcontrib/gpcloud/include/s3common_headers.h (hunk 3 above)
Comment (dnskvlnk) [reply gpcontrib/gpcloud/include/s3common_headers.h:36]: Can you control third-party code?

**Review by bandetto — APPROVED** [review]

File: arenadata/Dockerfile [hunk 4]
```diff
@@ -1,5 +1,14 @@
 FROM rockylinux:8.8 as base
 
+ARG CC=clang
```
Comment (andr-sokolov) [inline arenadata/Dockerfile:3]: The task is to fix Clang 18, 15 and 13 warnings, but Rocky Linux 8.8 has Clang 17. Can we use other docker image (with Clang 18) to check for the absence of Clang 18 warnings? This dockerfile may be useful for checking that GCC does not find warnings.

File: concourse/scripts/common.bash [hunk 5]
```diff
@@ -146,7 +146,16 @@ function configure() {
     # The full set of configure options which were used for building the
     # tree must be used here as well since the toplevel Makefile depends
     # on these options for deciding what to test. Since we don't ship
-    ./configure --prefix=/usr/local/greenplum-db-devel --disable-orca --enable-gpcloud --enable-orafce --enable-tap-tests --with-gssapi --with-libxml --with-openssl --with-perl --with-python --with-uuid=e2fs --with-llvm --with-zstd PYTHON=python3.11 PKG_CONFIG_PATH="${GPHOME}/lib/pkgconfig" ${CONFIGURE_FLAGS}
+
+	# temp to run tests on ci
+	export CC=clang
+	export CXX=clang++
+	if [ -z "${CC}" ] || [ -z "${CXX}" ]; then
```
Comment (andr-sokolov) [inline concourse/scripts/common.bash:153]: Why are these variables checked if they have been set in the previous line? Do we need to remove the export commands to use variables from ARG-s from arenadata/Dockerfile?

File: concourse/scripts/compile_gpdb.bash [hunk 6]
```diff
@@ -131,7 +131,6 @@ function export_gpdb_clients() {
 }
 
 function _main() {
-
```
Comment (andr-sokolov) [inline concourse/scripts/compile_gpdb.bash:134]: Is this change necessary to fix compiler warnings? I think it should be done in other patch

File: src/include/c.h [hunk 7]
```diff
@@ -1211,6 +1211,31 @@ typedef union PGAlignedXLogBlock
 	((underlying_type) (expr))
 #endif
 
+#if defined(__clang__) || (defined(__GNUC__) && __GNUC__ >= 7)
+	#define FALL_THROUGH __attribute__((fallthrough));
+#elif defined(__GNUC__)
+	#define FALL_THROUGH /* fall through */
```
Comment (andr-sokolov) [inline src/include/c.h:1217]: Did you test this case? Preprocessor removes all comments, so the warning is not fixed
```
[keremet@sl10 tmp]$ cat 1.c 
#define FT /* fall through */

int main(int argc, char**argv) {
	switch(argc){
	case 1:
		return 10;
	case 2:
		argc++;
		FT
	case 3:
		return 20;
	}
	return 0;
}
[keremet@sl10 tmp]$ gcc 1.c -Wimplicit-fallthrough=3
1.c: In function ‘main’:
1.c:8:7: warning: this statement may fall through [-Wimplicit-fallthrough=]
    8 |   argc++;
      |   ~~~~^~
1.c:10:2: note: here
   10 |  case 3:
      |  ^~~~
[keremet@sl10 tmp]$ gcc -E 1.c
# 1 "1.c"
# 1 "<built-in>"
# 1 "<command-line>"
# 31 "<command-line>"
# 1 "/usr/include/stdc-predef.h" 1 3 4
# 32 "<command-line>" 2
# 1 "1.c"


int main(int argc, char**argv) {
 switch(argc){
 case 1:
  return 10;
 case 2:
  argc++;
 
 case 3:
  return 20;
 }
 return 0;
}
[keremet@sl10 tmp]$ 
```

File: src/include/c.h [hunk 8]
```diff
@@ -1211,6 +1211,31 @@ typedef union PGAlignedXLogBlock
...
+#elif defined(_MSC_VER)
+	#define FALL_THROUGH [[fallthrough]];
+#endif
```
Comment (andr-sokolov) [inline src/include/c.h:1220]: When all the conditions are false, the FALL_THROUGH  macro is not defined and the program cannot be built. Do we need an empty default definition for FALL_THROUGH?

File: src/include/c.h [hunk 9]
```diff
@@ -1211,6 +1211,31 @@ typedef union PGAlignedXLogBlock
 	((underlying_type) (expr))
 #endif
 
+#if defined(__clang__) || (defined(__GNUC__) && __GNUC__ >= 7)
+	#define FALL_THROUGH __attribute__((fallthrough));
+#elif defined(__GNUC__)
+	#define FALL_THROUGH /* fall through */
+#elif defined(_MSC_VER)
```
Comment (andr-sokolov) [inline src/include/c.h:1218]: "Microsoft Visual C++ is a Windows-based tool" (https://learn.microsoft.com/en-us/answers/questions/1337459/is-there-a-msvc-support-on-linux-if-not-what-are-t). Greenplum does not support Windows, for example because there are no cgroups on Windows. Why is this case needed?

File: src/include/c.h [hunk 10]
```diff
@@ -1211,6 +1211,31 @@ typedef union PGAlignedXLogBlock
...
+		_Pragma("clang diagnostic pop") \
+	} while(0)
+#elif defined(__GNUC__)
+	#define SUPPRESS_COMPILER_WARNING(expr, warning) do { \
+		_Pragma("GCC diagnostic push") \
+		DO_PRAGMA(GCC diagnostic ignored warning) \
+		expr; \
+		_Pragma("GCC diagnostic pop") \
+	} while(0)
+#endif
```
Comment (andr-sokolov) [inline src/include/c.h:1237]: What about default definition for SUPPRESS_COMPILER_WARNING?

Comment (dnskvlnk) [reply src/include/c.h:1237]: Do we support anything other than gcc and clang? I believe that pragma syntax differs across compilers and we can't support all of them

File: src/include/c.h (hunk 9 above)
Comment (dnskvlnk) [reply src/include/c.h:1218]: I know that there's no msvc on linux and I can remove it, but would it harm if we ever going to move on vanilla postgres?

File: src/include/c.h (hunk 10 above)
Comment (andr-sokolov) [reply src/include/c.h:1237]: We don't use other compilers, but we should not prohibit to use them. The default definition may be something like `SUPPRESS_COMPILER_WARNING(expr, warning) expr`

File: arenadata/Dockerfile (hunk 4 above)
Comment (bandetto) [reply arenadata/Dockerfile:3]: I used to patch to compile GPDB on Clang 18 and 19 and got no warnings.

File: concourse/scripts/common.bash (hunk 5 above)
Comment (dnskvlnk) [reply concourse/scripts/common.bash:153]: CC and CXX were hardcoded because there's no configurable pipeline at the moment where we could choose a compiler, so the check is required. Once the review is done, these exports will be removed

File: arenadata/Dockerfile (hunk 4 above)
Comment (andr-sokolov) [reply arenadata/Dockerfile:3]: It is good, but my question is about future. Someday Clang 18 may find a warning that Clang 17 is not able to find. How can we detect this situation?

File: concourse/scripts/common.bash (hunk 5 above)
Comment (andr-sokolov) [reply concourse/scripts/common.bash:153]: If you want to remove exports, I suggest to replace `else` with `fi` because of `exit 1` and don't change the line with `./configure`

File: arenadata/Dockerfile [hunk 11]
```diff
@@ -1,5 +1,14 @@
...
+ARG CXX=clang++
+
+RUN /bin/bash -c $' \n\
+	if [ -z "${CC}" ] || [ -z "${CXX}" ]; then \n\
+		echo "User must provide CC and CXX" \n\
+		exit 1 \n\
+	fi'
```
Comment (andr-sokolov) [inline arenadata/Dockerfile:10]: Can we remove this step? The default values for CC and CXX are set. If user set CC and CXX to empty, then GCC is used.

**Comment by andr-sokolov** [comment]
What about creating a new dockerfile for clang? We need to check compilation only. It's redundant to run `make install` and mocking tests two times.

File: contrib/postgres_fdw/deparse.c [hunk 12]
```diff
@@ -3116,7 +3116,7 @@ deparsePartialAggFunctionParamFilter(Aggref *node, deparse_expr_cxt *context,
 			break;
 		case 2104:
 			/* float4 AVG function */
-			/* Fall through */
+			FALL_THROUGH
```
Comment (andr-sokolov) [inline contrib/postgres_fdw/deparse.c:3119]: Changes in contrib/postgres_fdw/deparse.c are redundant, because case follows case without any code between them

File: gpAux/Makefile [hunk 13]
```diff
@@ -96,8 +96,14 @@ SERVER_PLATFORMS=rhel7_x86_64 rhel6_x86_64 rhel8_x86_64 rocky8_x86_64 ol8_x86_64
...
-OPTFLAGS="$(strip $(BLD_CFLAGS) -O3 -fargument-noalias-global -fno-omit-frame-pointer -g)"
-PROFFLAGS="$(strip $(BLD_CFLAGS) -O3 -fargument-noalias-global -fno-omit-frame-pointer -g)"
+OPTFLAGS="$(strip $(BLD_CFLAGS) -O3 -fno-omit-frame-pointer -g)"
+PROFFLAGS="$(strip $(BLD_CFLAGS) -O3 -fno-omit-frame-pointer -g)"
+GCCFLAGS="-fargument-noalias-global"
+
+ifeq ($(firstword $(CC)), "gcc")
+OPTFLAGS=$(OPTFLAGS) $(GCCFLAGS)
+PROFFLAGS=$(PROFFLAGS) $(GCCFLAGS)
+endif
```
Comment (andr-sokolov) [inline gpAux/Makefile:106]: I can build the project without this change. What happens when it will be reverted?

**Comment by andr-sokolov** [comment]
The fix will be split into several PRs
